contracts/src/LightClient.sol

```diff
@@ -0,0 +1,152 @@
+    /// states are proposed.
+    error MissingLastBlockForCurrentEpoch(uint64 expectedBlockHeight);
+    /// @notice Invalid user inputs: wrong format or non-sensible arguments
+    error InvalidArgs();
+
+    constructor(LightClientState memory genesis, address stakeTableAddr) {
+        if (genesis.viewNum != 0 || genesis.blockHeight != 0) {
+            revert InvalidArgs();
+        }
+
+        genesisState = genesis;
+        currentEpoch = 0;
+        stakeTable = IStakeTable(stakeTableAddr);
+        // TODO: (alex) initialized stake table or at least store its contract address ref here
+    }
+
+    // === State Modifying APIs ===
+    //
+    /// @notice Update the latest finalized light client state. It is expected to be updated
+    /// periodically, especially an update for the last block for every epoch has to be submitted
+    /// before any newer state can be accepted since the stake table commitments of that block
+    /// become the snapshots used for vote verifications later on.
+    function newFinalizedState(
+        LightClientState calldata newState,
+        IPlonkVerifier.PlonkProof calldata proof
+    ) external {
+        if (newState.viewNum <= finalizedState.viewNum) {
+            revert OutdatedState();
+        }
+        uint64 epochEndingBlockHeight = (currentEpoch + 1) * BLOCKS_PER_EPOCH - 1;
+        if (
+            finalizedState.blockHeight != epochEndingBlockHeight
+                && newState.blockHeight > epochEndingBlockHeight
+        ) {
+            revert MissingLastBlockForCurrentEpoch(epochEndingBlockHeight);
+        }
+        // Sanity check (unnecessary under our threat model, just defense in depth)
+        if (newState.blockHeight <= finalizedState.blockHeight) {
+            revert InvalidArgs();
+        }
+        // format validity check
+        BN254.validateScalarField(newState.blockComm);
+        BN254.validateScalarField(newState.feeLedgerComm);
+        BN254.validateScalarField(newState.stakeTableBlsKeyComm);
+        BN254.validateScalarField(newState.stakeTableSchnorrKeyComm);
+        BN254.validateScalarField(newState.stakeTableAmountComm);
+
+        // check plonk proof
+        // TODO: (alex) replace the vk with the correct one
+        IPlonkVerifier.VerifyingKey memory vk = VkTest.getVk();
+        uint256[] memory publicInput = preparePublicInput(newState);
+        PlonkVerifier.verify(vk, publicInput, proof, bytes(""));
+
+        // upon successful verification, update state
+        if (newState.blockHeight == epochEndingBlockHeight) {
+            bytes32 newStakeTableComm = keccak256(
+                abi.encodePacked(
+                    newState.stakeTableBlsKeyComm,
+                    newState.stakeTableSchnorrKeyComm,
+                    newState.stakeTableAmountComm
+                )
+            );
+
+            stakeTable.advanceEpoch(newStakeTableComm);
+        }
+        if (finalizedState.blockHeight == epochEndingBlockHeight) {
```

Comment:
```suggestion
        if (newState.blockHeight == epochEndingBlockHeight) {
```

Because `finalizedState` is updated afterwards only? Or you can merge with the if statement at line 114.

Reply:
no, (but you are right that my logic is also wrong):
Case 1: `newState` is in the middle of an epoch, --> just update `finalizedState`
Case 2: `newState` is the last block of an epoch --> `epoch` should _NOT_ increment, still just update `finalizedState`
Case 3: `newState` is in the next epoch --> `epoch` should increment, and `stakeTable` should advance with the last stored `finalizedState`'s commitment (cuz that was the state of the last block of the last epoch), only then update `finalizedState`